sw/host/opentitanlib/src/transport/qemu/subprocess.rs

```diff
@@ -0,0 +1,235 @@
+                s
+            });
+        command.args(["-machine", &machine]);
+
+        // Add arguments for specified images:
+
+        if let Some(rom_image) = options.rom_image {
+            command.arg("-kernel");
+            command.arg(rom_image);
+        }
+
+        // NOTE: We load these drives as `readonly` because Bazel can
+        // only provide readonly test inputs. To support tests which modify
+        // flash or OTP, we will need to make writable copies of the images
+        // ourselves.
+
+        if let Some(flash_image) = options.flash_image {
+            command.arg("-drive");
+            command.arg(format!(
+                "if=mtd,bus=1,file={},format=raw,readonly=on",
+                flash_image.display()
+            ));
+        }
+
+        if let Some(otp_image) = options.otp_image {
+            command.arg("-drive");
+            command.arg(format!(
+                "if=pflash,file={},format=raw,readonly=on",
+                otp_image.display()
+            ));
+        }
+
+        // Create files for QEMU to connect IO to.
+        let socket_dir = tempfile::tempdir().context("failed to create tempdir")?;
+
+        let uart_path = socket_dir.path().join(Self::UART0_SOCKET);
+        command.arg("-chardev");
+        command.arg(format!(
+            "socket,id=uart0,server=on,path={}",
+            uart_path.display()
+        ));
+        command.args(["-serial", "chardev:uart0"]);
+
+        // Add explicitly requested arguments.
+        command.args(options.extra_args.as_slice());
+
+        log::info!("CWD: {:?}", env::current_dir());
+        log::info!(
+            "Spawning QEMU: {} {}",
+            options.executable.display(),
+            command
+                .get_args()
+                .map(|s| s.to_string_lossy())
+                .collect::<Vec<_>>()
+                .join(" ")
+        );
+
+        // Start QEMU.
+        let mut child = command
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()?;
+
+        // Take its stderr and accumulate it in a string.
+
+        let stderr = child
+            .stderr
+            .take()
+            .context("failed to take process stderr")?;
+
+        let accumulated_stderr: Arc<Mutex<String>> = Default::default();
+        let accumulator = Arc::clone(&accumulated_stderr);
+
+        thread::spawn(move || {
+            printer::accumulate(stderr, concat!(module_path!(), "::stderr"), accumulator)
+        });
+
+        Ok(Subprocess {
+            child,
+            socket_dir,
+            accumulated_stderr,
+        })
+    }
+
+    pub fn uart_path(&self) -> PathBuf {
+        self.socket_dir.path().join(Self::UART0_SOCKET)
+    }
+
+    /// Finds a string within the QEMU output.
+    /// It is assumed that the [`Regex`] `re` has zero or one capture groups.
+    pub fn find(&self, re: &Regex, deadline: Instant) -> anyhow::Result<String> {
```

Comment:
I am struggling to see the flow here: this function is called after creating the subprocess. But the subprocess constructor already spawns a thread to print the accumulated stderr. Isn't there a risk that the printer will eat the message you are looking for before this function has the time to lock `self.accumulated_stderr`?

Reply:
I think the `printer` is just redirecting `stderr` to the `Subprocess::accumulated_stderr` string which is what `fn find` is reading from, so they aren't both trying to read from the same source.

I'm not very happy with this function but copied it from the verilator module (maybe should have moved it somewhere to share)